### PR TITLE
Add graph to author analytics page

### PR DESCRIPTION
**Pull request by Will-Howard**
See designs [here](https://www.figma.com/file/d09iVzNSpZp9GyWGcfkWmX/Q2-2023?type=design&node-id=1335-88028&mode=design&t=vUy89FthW4E7Nxtd-0), although there wasn't really a design for the graph so this was largely improvised.

This adds an `AnalyticsGraph` component and hooks/resolvers to supply it. This is used in the author analytics page, and will be used in the post analytics page. I've only added it in the author analytics page so far, I plan to do the post analytics page all in one go.

Things to note:
 - I had a few more problems with the queries being slow, again I think I've improved them all to a reasonable level by adding indexes etc. Still though based on this I decided to refactor both the post list and the graph to fetch the materialised data first, and then do a followup query to fetch any live data that is missing. This is a bit of a shame as it adds a lot of complexity to the hooks in useAnalytics, but it does make it so the page loads in a couple of seconds for all users (even ones who have tons of posts)

Screenshot 2023-08-14 at 15 20 55
Screenshot 2023-08-14 at 15 21 14
Screenshot 2023-08-14 at 15 22 52

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1205250724125315) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/server/resolvers/analyticsResolvers.ts [hunk 1]
```diff
@@ -177,27 +195,23 @@ addGraphQLResolvers({
       // keys, the partial is no longer partial
       return postAnalytics as PostAnalyticsResult;
     },
+    // TODO make this handle posts as well
```
Comment (oetherington) [inline packages/lesswrong/server/resolvers/analyticsResolvers.ts:198]: I'm assuming this TODO is meant to still be here

File: packages/lesswrong/server/resolvers/analyticsResolvers.ts [hunk 2]
```diff
@@ -357,6 +342,146 @@ addGraphQLResolvers({
...
+        (new VotesRepo().getPostKarmaChangePerDay({
+          postIds: queryPostIds,
+          startDate: adjustedStartDate?.toDate(),
+          endDate: adjustedEndDate.toDate(),
+        })),
+        (new CommentsRepo().getCommentsPerDay({
+          postIds: queryPostIds,
+          startDate: adjustedStartDate?.toDate(),
+          endDate: adjustedEndDate.toDate(),
+        })),
```
Comment (oetherington) [inline packages/lesswrong/server/resolvers/analyticsResolvers.ts:454]:
```suggestion
        context.repos.votes.getPostKarmaChangePerDay({
          postIds: queryPostIds,
          startDate: adjustedStartDate?.toDate(),
          endDate: adjustedEndDate.toDate(),
        }),
        context.repos.comments.getCommentsPerDay({
          postIds: queryPostIds,
          startDate: adjustedStartDate?.toDate(),
          endDate: adjustedEndDate.toDate(),
        }),
```

**Review by oetherington — APPROVED** [review]
Looks really good! I left a few comments, but nothing major.

File: packages/lesswrong/components/hooks/useAnalytics.ts [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+      }
+    }
+  `;
+  const { userId, postIds, startDate: propsStartDate, endDate: propsEndDate } = props;
+
+  const fetchDateRef = useRef({ startDate: propsStartDate, endDate: propsEndDate });
+  
+  if (
+    (fetchDateRef.current.startDate ?? 0) > (propsStartDate ?? 0) ||
+    // TODO make clearer
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/components/hooks/useAnalytics.ts:192]: No, I've replaced it with a proper comment

File: packages/lesswrong/components/analytics/AnalyticsGraph.tsx [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+  },
+  dateDropdown: {
+    alignSelf: "flex-start",
+    margin: '4px 0'
+  }
+});
+
+const LINE_COLORS: Record<AnalyticsField, string> = {
+  reads: requireCssVar("palette", "primary", "main"),
+  views: requireCssVar("palette", "primary", "light"),
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/components/analytics/AnalyticsGraph.tsx:131]: Good point, I've swapped the karma and reads colours. There is still a problem when showing views and karma together but this should be better as that's not the default view, and the numbers will tend to be further apart which distinguishes them more
Screenshot 2023-08-18 at 12 18 48

File: packages/lesswrong/server/resolvers/analyticsResolvers.ts (hunk 1 above)
Comment (Will-Howard) [reply packages/lesswrong/server/resolvers/analyticsResolvers.ts:198]: Yeah it is, I'll reuse this for the post analytics page

**Pull request merged**